File: JavaScript/Arrays & Hashing/Medium/design-linked-list.js
/* Question Link:  https://leetcode.com/problems/design-linked-list/ */

var MyLinkedList = function() {
    this.size = 0;
    this.head = {val: 0, next: null};
};

/** 
 * @param {number} index
 * @return {number}
 */
MyLinkedList.prototype.get = function(index) {
   if(index < 0 || index >= this.size) return -1;

   let curr = this.head.next;
   for(let i = 0; i < index; i++){
       curr = curr.next;
   }

   return curr.val;
};

/** 
 * @param {number} val
 * @return {void}
 */
MyLinkedList.prototype.addAtHead = function(val) {
    this.addAtIndex(0, val);
};

/** 
 * @param {number} val
 * @return {void}
 */
MyLinkedList.prototype.addAtTail = function(val) {
    this.addAtIndex(this.size, val);
};

/** 
 * @param {number} index 
 * @param {number} val
 * @return {void}
 */
MyLinkedList.prototype.addAtIndex = function(index, val) {
    if(index < 0) index = 0;
    if(index > this.size) return;

    let prev = this.head;
    for(let i = 0; i < index; i++){
        prev = prev.next;
    }

    const node = {val: val, next: prev.next};
    prev.next = node;
    this.size++;
};

/** 
 * @param {number} index
 * @return {void}
 */
MyLinkedList.prototype.deleteAtIndex = function(index) {
    if(index < 0 || index >= this.size) return;

    let prev = this.head;
    for(let i = 0; i < index; i++){
        prev = prev.next;
    }
    prev.next = prev.next.next;
    this.size--;
};

/** 
 * Your MyLinkedList object will be instantiated and called as such:
 * var obj = new MyLinkedList()
 * var param_1 = obj.get(index)
 * obj.addAtHead(val)
 * obj.addAtTail(val)
 * obj.addAtIndex(index,val)
 * obj.deleteAtIndex(index)
 */

File: LeetCode/Python/Linked List/Medium/design-linked-list.py
# Question Link: https://leetcode.com/problems/design-linked-list/

class MyLinkedList(object):

    def __init__(self):
        self.head = Node()
        self.size = 0

    def get(self, index):
        """
        :type index: int
        :rtype: int
        """
        if index < 0 or index >= self.size:
            return -1
        curr = self.head.next
        for _ in range(index):
            curr = curr.next
        return curr.val
        

    def addAtHead(self, val):
        """
        :type val: int
        :rtype: None
        """

        node = Node(val)
        node.next = self.head.next
        self.head.next = node
        self.size += 1
        

    def addAtTail(self, val):
        """
        :type val: int
        :rtype: None
        """

        curr = self.head
        while curr.next:
            curr = curr.next
        curr.next = Node(val)
        self.size += 1
        

    def addAtIndex(self, index, val):
        """
        :type index: int
        :type val: int
        :rtype: None
        """

        if index < 0 or index > self.size:
            return
        curr = self.head
        for _ in range(index):
            curr = curr.next
        node Node(val)
        node.next = curr.next
        curr.next = node
        self.size += 1
        

    def deleteAtIndex(self, index):
        """
        :type index: int
        :rtype: None
        """
        


# Your MyLinkedList object will be instantiated and called as such:
# obj = MyLinkedList()
# param_1 = obj.get(index)
# obj.addAtHead(val)
# obj.addAtTail(val)
# obj.addAtIndex(index,val)
# obj.deleteAtIndex(index)
